inventory/base_inventory.py

```diff
@@ -0,0 +1,99 @@
+#!/usr/bin/env python
+# Copyright 2018, Rackspace US, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import argparse
+import copy
+import json
+import os
+
+
+class MaasInventory(object):
+
+    def __init__(self):
+        self.inventory = self.empty_inventory()
+        self.read_cli_args()
+
+        # main logic, generate inventory dynamically
+        self.generate_inventory()
+
+        # Called with `--list`.
+        if self.args.list:
+            data = self.inventory
+        # Called with `--host [hostname]`.
+        elif self.args.host:
+            # Not needed, since we return `_meta` with `--list`.
+            data = self.host_inventory(self.args.host)
+        # If no groups or vars are present, return an empty inventory.
+        else:
+            data = self.empty_inventory()
+
+        data = json.dumps(data)
+
+        if self.args.outfile:
+            self._write_to_file(data, self.args.outfile)
+        else:
+            print(data)
+
+    def _write_to_file(self, data, path):
+        if os.access(os.path.dirname(path), os.W_OK):
+            with open(path, 'w') as outfile:
+                outfile.write(data)
+        else:
+            raise RuntimeError("FATAL: Unable to write to {path}".format(
+                path=path))
+
+    def generate_env_specific_variables(self):
```

Comment:
Hey @briancurtin I am taking this over for Tony while he is gone to help move things a long. Can you could explain why implementing as an `ABC` class would be better?

Reply:
`NotImplementedError` is more of a thing you use while code is in-development. `ABC` solves the purpose of a class being a real abstract base class. 

You don't want an unimplemented method on a subclass sneaking in and blowing up at runtime for a user with `NotImplementedError`. For example, that could have occurred here with the typo if it only existed in one place. You would want your tests to not even run at all because you have a class that doesn't implement the proper interface.